### DISCUSSION TITLE
Content security policy options are not working

### DISCUSSION THREAD

**Discussion by FINDarkside** [directus/directus#15865] [Archive]
### Describe the Bug

None of the env variables used to change CSP actually seem to work. By default our directus responds with `Content-Security-Policy: default-src none` which means that our safari users aren't able to open pdf files hosted in directus.

See: #12420 except that I have verified that the same happens even with no reverse proxies, running on localhost. 

In that issue it was also mentioned that Directus isn't supposed to default to `Content-Security-Policy: default-src none` but I can reproduce `default-src none` even with free community cloud instance. See https://3aa46c29.directus.app/assets/a3df3832-89c7-487f-9c36-8bac1c6c499f

### To Reproduce

Just copy all the CSP's from #12420 
```
CONTENT_SECURITY_POLICY: "media-src 'self' https://patreon.com https://bray.tech https://beta.bray.tech"
CONTENT_SECURITY_POLICY_DIRECTIVES__MEDIA_SRC: "media-src 'self' https://patreon.com https://bray.tech https://beta.bray.tech"
CONTENT_SECURITY_POLICY: "{ directives: { mediaSrc: ['self', 'https://patreon.com', 'https://bray.tech', 'https://beta.bray.tech'], }, }"
ASSETS_CONTENT_SECURITY_POLICY: "media-src 'self' https://patreon.com https://bray.tech https://beta.bray.tech"
```
And open directus file assets url and I still get `Content-Security-Policy: default-src none`

**Comment by azrikahar** [comment]
The `ASSETS_CONTENT_SECURITY_POLICY` entry from that issue was actually incorrect, but understandable it should've been mentioned/corrected nonetheless.

Setting it as `ASSETS_CONTENT_SECURITY_POLICY_DIRECTIVES__MEDIA_SRC="array:'self',https://patreon.com,https://bray.tech,https://beta.bray.tech"` does work:

![chrome_ai3XSm6JRU](https://user-images.githubusercontent.com/42867097/193558443-03c11841-c2f5-43cb-8929-fde3070a6ff4.png)

The assets endpoints defaults to zero trust as noted in the PR #12020, which is why the `/assets` defaults to `default-src none` and requires explicit configuration of `ASSETS_CONTENT_SECURITY_POLICY` environment variable(s).

**Reply by FINDarkside** [reply]
Okay so `ASSETS_CONTENT_SECURITY_POLICY_DIRECTIVES__MEDIA_SRC` is the correct env var, the value itself was just fine I believe, I'm not sure where that array syntax comes from but normal CSP syntax seems to work just fine. Why is `ASSETS_CONTENT_SECURITY_POLICY` not working though and how am I supposed to know these env variables? The docs link to front page of Helmet which as far as I can see do not mention anything like `ASSETS_CONTENT_SECURITY_POLICY_DIRECTIVES__MEDIA_SRC` anywhere.

Seems like I can use `ASSETS_CONTENT_SECURITY_POLICY_DIRECTIVES__DEFAULT_SRC` to change the default src as well so this is enough for my usecase. But I don't think the docs give enough information to actually use this feature at the moment.

**Reply by azrikahar** [reply]
> I'm not sure where that array syntax comes from

It is one of the supported prefix for type casting: https://docs.directus.io/self-hosted/config-options.html#environment-syntax-prefix 👍 Technically yes it can be omitted as it will be auto typecasted as of now, but there has been some confusion with the "auto type cast" itself (and there are times where users _do not_ want it to by auto casted), hence the expicit example showing `array:` in it. 

Also note that `ASSETS_CONTENT_SECURITY_POLICY_DIRECTIVES__DEFAULT_SRC='self'` (without array prefix) will be interpreted as a string and result in `default-src self` instead of `default-src 'self'` (with quotes), which isn't a valid value according to CSP specs AFAIK.

This is also being explored in #9521 as a breaking change in v10, where prefixing it as an array may be compulsory in future versions.

> Why is `ASSETS_CONTENT_SECURITY_POLICY` not working though and how am I supposed to know these env variables? 

This is because it will need to "mirror" helmet's config. If we were to base it on one of helmet's example:

![chrome_AA9esviCsR](https://user-images.githubusercontent.com/42867097/193562641-c1d54b1b-ce36-4eba-93e3-4700175c4488.png)

the "translation" to Directus-specific environment variables would be:

```yaml
ASSETS_CONTENT_SECURITY_POLICY_USE_DEFAULT=false
ASSETS_CONTENT_SECURITY_POLICY_DIRECTIVES__DEFAULT_SRC=array:'self'
ASSETS_CONTENT_SECURITY_POLICY_DIRECTIVES__SCRIPT_SRC=array:'self',example.com
ASSETS_CONTENT_SECURITY_POLICY_DIRECTIVES__OBJECT_SRC=array:'none'
ASSETS_CONTENT_SECURITY_POLICY_DIRECTIVES__UPGRADE_INSECURE_REQUESTS=[]
```

> The docs link to front page of Helmet which as far as I can see do not mention anything like `ASSETS_CONTENT_SECURITY_POLICY_DIRECTIVES__MEDIA_SRC` anywhere.

The docs link is meant to point to the `contentSecurityPolicy` options of helmetjs which requires scrolling down and expanding the detail (which understandably isn't as obvious as it should be):

![image](https://user-images.githubusercontent.com/42867097/193565587-77f4d5a6-3169-4016-ae1e-660189e7aed0.png)

The double underscore syntax to nest variables is mentioned over here: https://docs.directus.io/self-hosted/config-options.html#type-casting-and-nesting

---

With that said, we're actively working on the docs, so any suggestions to improve this topic are very much appreciated!

**Comment by cliqer** [comment]
A quick fix to be able to load your assets is to add:
```
ASSETS_CONTENT_SECURITY_POLICY_DIRECTIVES__DEFAULT_SRC=array:'unsafe-inline'
```
inside .env file in directus

**Reply by ricricucit** [reply]
⚠️ ...while lowering your security level by quite a lot!

**Comment by mikgross** [comment]
I am getting a 403 forbidden on a public supabase file storage with directus.
The documentation is highly unclear on how to format the env variables to configure helmet. It would be VERY nice to add examples on how to configure the assets content access policies.

**Comment by vinduajamesh** [comment]
This is really an annoying problem not only for me but for our clients and users. Please update your documentation on self hosting on how to address this.

**Comment by slavaWins** [comment]
Я определил как называется env параметр:
      CONTENT_SECURITY_POLICY_DIRECTIVES__imgSrc: "array:'self',blob:,https://*"


Если оставить как было в конфиге + своё добавить:
      CONTENT_SECURITY_POLICY_DIRECTIVES__imgSrc: "array:'self',data:,blob:,https://raw.githubusercontent.com,https://avatars.githubusercontent.com,https://*"
      


На случай если настройки поменяются, оставлю инструкцию как найти название параметра
Идем сюда:
https://github.com/directus/directus/blob/main/api/src/app.ts


Там есть штука которая мерджит енв в настройки приватности:
Вот по этому принципу можно подобрать путь
	getConfigFromEnv('CONTENT_SECURITY_POLICY_'), + imgSrc 




```
app.use(
		helmet.contentSecurityPolicy(
			merge(
				{
					useDefaults: true,
					directives: {
						// Unsafe-eval is required for app extensions
						scriptSrc: ["'self'", "'unsafe-eval'"],

						// Even though this is recommended to have enabled, it breaks most local
						// installations. Making this opt-in rather than opt-out is a little more
						// friendly. Ref #10806
						upgradeInsecureRequests: null,

						// These are required for MapLibre
						workerSrc: ["'self'", 'blob:'],
						childSrc: ["'self'", 'blob:'],
						imgSrc: [
							"'self'",
							'data:',
							'blob:',
							'https://raw.githubusercontent.com',
							'https://avatars.githubusercontent.com',
						],
						mediaSrc: ["'self'"],
						connectSrc: ["'self'", 'https://*', 'wss://*'],
					},
				},
				getConfigFromEnv('CONTENT_SECURITY_POLICY_'),
			),
